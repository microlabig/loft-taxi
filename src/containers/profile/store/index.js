export { default } from './reducer';

export * from './actions';
//export * from './middlewares';
export * from './sagas';
export * from './selectors';