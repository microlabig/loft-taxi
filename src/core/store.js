import { createStore, combineReducers, compose, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { all } from 'redux-saga/effects';

import userReducer, { userSaga } from '../containers/login/store';
import cardReducer, { cardSaga } from '../containers/profile/store';
import addressReducer, { addressSaga } from '../containers/map/store';

// делаем общий срез редьюсеров
export const rootReducer = combineReducers({
    userReducer,
    cardReducer,
    addressReducer
});

// создаем сагу 
const sagaMiddleware = createSagaMiddleware();

// создаем root-сагу для запуска отдельных саг параллельно
function* rootSaga() {
    yield all([
        userSaga(),
        cardSaga(),
        addressSaga()
    ]);
}

// создаем стор
export default createStore(
    rootReducer,
    // объединяем функции справа-налево
    compose(
        // добавляем созданную выше сагу
        applyMiddleware(sagaMiddleware),
        // подключаем расширение Redux DevTools для дебага redux
        window.__REDUX_DEVTOOLS_EXTENSION__
            ? window.__REDUX_DEVTOOLS_EXTENSION__()
            : noop => noop // для тестов
    )
); 

// запускаем root-сагу
sagaMiddleware.run(rootSaga);
