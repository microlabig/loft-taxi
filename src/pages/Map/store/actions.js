import { createAction } from 'redux-actions';
import * as consts from './consts';

// Route
export const fetchRouteLoading = createAction(consts.ROUTE_LOADING);
export const fetchRouteRequest = createAction(consts.ROUTE_REQUEST);
export const fetchRouteSuccess = createAction(consts.ROUTE_SUCCESS);
export const fetchRouteFailure = createAction(consts.ROUTE_FAILURE);

// AddressList
export const fetchAddressLoading = createAction(consts.ADDRESS_LOADING);
export const fetchAddressListRequest = createAction(consts.ADDRESS_LIST_REQUEST);
export const fetchAddressListSuccess = createAction(consts.ADDRESS_LIST_SUCCESS);
export const fetchAddressListFailure = createAction(consts.ADDRESS_LIST_FAILURE);

export const fetchAddressClearError = createAction(consts.ADDRESS_CLEAR_ERROR);
export const fetchAddressReset = createAction(consts.ADDRESS_RESET);