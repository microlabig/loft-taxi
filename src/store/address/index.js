export { default } from './reducer';

export * from './actions';
export * from './middlewares';
export * from './selectors';