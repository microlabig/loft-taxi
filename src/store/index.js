import { createStore, combineReducers, compose, applyMiddleware } from 'redux';

import user, { userMiddleware } from './user';
import card, { cardMiddleware } from './card';

export default createStore(
    combineReducers({
        user,
        card
    }),
    compose(
        applyMiddleware(userMiddleware),
        applyMiddleware(cardMiddleware),
        window.__REDUX_DEVTOOLS_EXTENSION__
            ? window.__REDUX_DEVTOOLS_EXTENSION__
            : noop => noop
    )
); 