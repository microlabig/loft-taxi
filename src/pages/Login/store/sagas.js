import { takeLatest, call, put, fork } from 'redux-saga/effects';
import * as consts from './consts';
import * as actions from './actions';
import * as api from '../../../utils/api';

// сага обработки промежуточных результатов 
function* fetchUserWorker(action) {
    const { type, payload } = action;
    let apiFunc = null;

    // тип экшена откуда произошел диспатч - логин или регистрация
    // для настройки передачи ф-ии api
    switch (type) {
        // логин
        case consts.USER_LOGIN:
            apiFunc = api.userLogin;
            break;
        // регистрация
        case consts.USER_REGISTER:
            apiFunc = api.userRegister;
            break;
        default:
            return;
    }

    // диспатчим экшен установки флага начало загрузки инф-ии
    yield put(actions.fetchUserLoading());

    try {
        // вызываем ф-ию api и ждем когда прийдет ответ от сервера
        const response = yield call(apiFunc, payload);
        // проверяем статус ответа
        if (response.success) {
            // диспатчим экшен успешного запроса
            yield put(actions.fetchUserSuccess(response));
        } else {
            throw new Error(response.error);
        }
    } catch (error) {
        // диспатчим экшен в случае ошибки запроса
        yield put(actions.fetchUserFailure(error.message));
    }
}

// сага возврата последнего вызова fetchUserWorker в стеке запросов логина пользователя
function* loginWatcher() {
    yield takeLatest(consts.USER_LOGIN, fetchUserWorker);
}

// сага возврата последнего вызова fetchUserWorker в стеке запросов регистрации пользователя
function* registerWatcher() {
    yield takeLatest(consts.USER_REGISTER, fetchUserWorker);
}

// общая сага user data запуска неблокирующих запросов
export function* userSaga() {
    yield fork(loginWatcher, null);
    yield fork(registerWatcher, null);
}