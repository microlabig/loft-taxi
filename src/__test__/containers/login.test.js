import React from 'react';
import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";

import { Provider } from 'react-redux'
import { createStore } from 'redux';
import { rootReducer } from "../../store";

import LoginPage from '../../containers/Login';
import FormLogin from '../../containers/Login/formLogin';

// Сделаем мок-компонент Map
jest.mock('mapbox-gl/dist/mapbox-gl', () => ({
    Map: () => ({}),
}));

describe('компонент Login', () => {
    let initStore = null;
    let wrapper = null;

    // подготавливаем DOM-элемент, куда будем рендерить
    beforeEach(() => { // ..Each - выполняется перед каждой ф-ией it
        initStore = createStore(rootReducer);
        wrapper = mount(
            <MemoryRouter>
                <Provider store={initStore}>
                    <LoginPage />
                </Provider>
            </MemoryRouter>
        );
    });

    // подчищаем после завершения
    afterEach(() => { // ..Each - выполняется перед каждой ф-ией it
        wrapper.unmount();
        initStore = null;
        wrapper = null;
    });

    describe('инициализация', () => {
        it('монтирование компонента', () => {
            expect(wrapper.find(LoginPage)).toHaveLength(1);
        });

        it('есть необходимая текстовая информация', () => {
            expect(wrapper.find('h1').text()).toEqual("Loft Taxi");
            expect(wrapper.find('h2').text()).toEqual("Войти");
        });

        it('есть ссылка для перехода на регистрацию пользователя', () => {
            expect(wrapper.find('a').text()).toEqual("Зарегистрируйтесь");
        });

        it('есть поля ввода', () => {
            expect(wrapper.find('input')).toHaveLength(2);
        });

        it('есть кнопка входа', () => {
            expect(wrapper.find('button')).toHaveLength(1);
        });
    });

    describe('проверка состояния элементов компонента', () => {
        let values = null;
        // подготавливаем DOM-элемент, куда будем рендерить
        beforeEach(() => { // ..Each - выполняется перед каждой ф-ией it
            values = {
                email: 'name@domain.com', password: 'password123'
            }
        });

        // подчищаем после завершения
        afterEach(() => { // ..Each - выполняется перед каждой ф-ией it
            values = null;
        });

        it('создание снапшота', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('при монтировании компонента кнопка неактивна', () => {
            expect(wrapper.find('button').prop('disabled')).toBeTruthy();
        });

        it('при вводе данных в разные поля ввода кнопка неактивна', async () => {
            const emailInput = wrapper.find('[data-testid="input-email"]');
            const passwordInput = wrapper.find('[data-testid="input-password"]');

            await act(async () => {
                await emailInput.simulate('change', {
                    target: {
                        name: 'email',
                        value: values.email
                    }
                });
            });

            expect(wrapper.find('button').prop('disabled')).toBeTruthy();

            await act(async () => {
                await emailInput.simulate('change', {
                    target: {
                        name: 'email',
                        value: ''
                    }
                });
                await passwordInput.simulate('change', {
                    target: {
                        name: 'password',
                        value: values.password
                    }
                });
            });

            expect(wrapper.find('button').prop('disabled')).toBeTruthy();
        });

        it('при вводе данных во все поля ввода кнопка активна', async () => {
            const emailInput = wrapper.find('[data-testid="input-email"]');
            const passwordInput = wrapper.find('[data-testid="input-password"]');

            await act(async () => {
                await emailInput.simulate('change', {
                    target: {
                        name: 'email',
                        value: values.email
                    }
                });
                await passwordInput.simulate('change', {
                    target: {
                        name: 'password',
                        value: values.password
                    }
                });
            });

            expect(wrapper.find('button').prop('disabled')).toBeFalsy();
        });

        /* it('форма сабмитится', async () => {
            //const instance = wrapper.instance();
            const form = wrapper.find('form');
            const spy = jest.spyOn(form.props(), 'onSubmit');

            wrapper.update();
            wrapper.instance().forceUpdate();

            const emailInput = wrapper.find('[data-testid="input-email"]');
            const passwordInput = wrapper.find('[data-testid="input-password"]');
            const button = wrapper.find('button');
            
            await act(async () => {
                await emailInput.simulate('change', {
                    target: {
                        name: 'email',
                        value: values.email
                    }
                });
                await passwordInput.simulate('change', {
                    target: {
                        name: 'password',
                        value: values.password
                    }
                });
                await button.simulate('click',{ target: { name: 'submit' } });
                //await form.prop('onSubmit')();
            });

            expect(spy).toHaveBeenCalled();
            spy.mockClear();
        }); */
    })
});