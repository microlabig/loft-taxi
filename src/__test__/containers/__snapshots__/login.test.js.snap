// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`компонент Login проверка состояния элементов компонента создание снапшота 1`] = `ReactWrapper {}`;
